### ISSUE TITLE
BeaconState::get_active_validator_indices should check epoch

### ISSUE THREAD

**Issue by michaelsproul** [sigp/lighthouse#922]
## Description

The function `BeaconState::get_active_validator_indices` takes an arbitrary epoch and attempts to compute (without the committee cache) the active validators at that epoch. The problem is, we only have limited lookahead because of new validators activating and old validators leaving.

## Steps to resolve

1. Work out the minimum lookahead on the active validator set by reasoning about the activation queue, finalization, and the exit queue. Probably the most relevant part of the spec is [`process_registry_updates`](https://github.com/ethereum/eth2.0-specs/blob/v0.10.1/specs/phase0/beacon-chain.md#registry-updates). Based on some back of the envelope math it looks like maybe the safe lookahead is 2 epochs, i.e. 3 epoch transitions from when a deposit is made: 1st adds validator to the activation queue with eligibility in the next epoch, 2nd justifies the eligible epoch (best case), 3rd finalizes the eligible epoch (best case) thus activating the validator.
2. Add a check on the epoch to `get_active_validator_indices`, returning an error when it is too high (or indeed too low).

Alternatively, look into whether `get_active_validator_indices` actually needs to be accessible from the `BeaconState`. It's used by `get_beacon_proposer_index` which requires a sorted list of validator indices, and therefore can't use the active indices from the committee cache.

**Comment by ethDreamer** [comment]
I will take a look at this.

**Comment by ethDreamer** [comment]
@michaelsproul 
I'm a little confused about this question after looking into it for a while. You say:
> work out the minimum lookahead
> return an error [when epoch is] too low

Correct me if I'm wrong but the active validator set is known for the current epoch and every past epoch correct? So I'm not sure what you mean by minimum lookahead / too low (other than a negative number).

I've had a bunch of stuff come up recently so I haven't had much time to look at this. I believe your maximum of 3 epoch transitions is correct at least as far as the activation queue goes but I still need to work out the exit queue.

**Comment by michaelsproul** [comment]
@ethDreamer You're right! There's no lower limit! Thanks for looking into this :)

**Comment by ethDreamer** [comment]
Okay so I'll try to write this out in steps simplifying as I go:

There are 4 events that could change the active validator set:

1. NEW DEPOSIT:
```
state_transition():
	process_slots():
		process_epoch():  // once per epoch when (state.slot + 1) % SLOTS_PER_EPOCH == 0
	process_block():
		process_operations():
			process_deposit(): // leads to one of the following which can make validator eligible for activation queue
				1:	get_validator_from_deposit()
				2:	increase_balance()
```
2. VALIDATOR SLASHED
```
state_transition()
	process_slots():
		process_epoch():  // once per epoch when (state.slot + 1) % SLOTS_PER_EPOCH == 0
	process_block():
		process_operations():
			process_proposer_slashing() / process_attester_slashing():
				slash_validator():
					initiate_validator_exit():
						validator.exit_queue_epoch = compute_activation_exit_epoch(); // MINIMUM
```
3. VOLUNTARY EXIT
```
state_transition()
	process_slots():
		process_epoch():  // once per epoch when (state.slot + 1) % SLOTS_PER_EPOCH == 0
	process_block():
		process_operations():
			process_voluntary_exit():
				initiate_validator_exit():
					validator.exit_queue_epoch = compute_activation_exit_epoch(); // MINIMUM
```
4. PENALTIES FORCE VALIDATOR TO EXIT
```
state_transition():
	process_slots():
		process_epoch(): // once per epoch when (state.slot + 1) % SLOTS_PER_EPOCH == 0
			...
			process_rewards_and_penalties():
			// expanded below
```

I need to expand on process_epoch():
```
process_epoch(): // once per epoch when (state.slot + 1) % SLOTS_PER_EPOCH == 0
	process_justification_and_finalization(state)
		// skimming this code I *think* (don't quote me) the smallest epoch increments are:
		JUSTIFICATION_GAP = 1
		FINALIZATION_GAP = 2
	process_rewards_and_penalties(state)
	process_registry_updates(state)
		foreach validator:
			if eligible_for_activation_queue():
				validator.activation_eligibility_epoch = current_epoch+1
			if validator.effective_balance <= EJECTION_BALANCE
				initiate_validator_exit():  // this is case 4
					validator.exit_queue_epoch = compute_activation_exit_epoch(); // MINIMUM
		foreach validator in activation_queue:
			if validator.activation_eligibility_epoch <= state.finalized_checkpoint.epoch
				validator.activation_epoch = compute_activation_exit_epoch(); // MINIMUM
	process_slashings(state)
```

For case 1, due to the order of process_slots() THEN process_block(), notice that if the current slot satisfies `(state.slot + 1) % SLOTS_PER_EPOCH == 0`, then we need to add one extra epoch to the answer.

I will simplify by noting that:
```
compute_activation_exit_epoch(epoch):
	return epoch + 1 + MAX_SEED_LOOKAHEAD // = epoch + 1 + 4 = epoch + 5
```
Therefore:
1. NEW DEPOSIT
```
// assume deposit was made and validator added this epoch
validator.activation_epoch = compute_activation_exit_epoch(FINALIZATION_GAP + (current_epoch + 1))
```
2-4. VALIDATOR SLASHED / VOLUNTARY EXIT / PENALTIES FORCE VALIDATOR EXIT:
```
validator.exit_queue_epoch = compute_activation_exit_epoch(current_epoch); // MINIMUM	
```
simplifies to:
1. NEW DEPOSIT
```
validator.activation_epoch = current_epoch + FINALIZATION_GAP + 6
```
2-4. VALIDATOR SLASHED / VOLUNTARY EXIT / PENALTIES FORCE VALIDATOR EXIT:
```
validator.exit_queue_epoch = current_epoch + 5
```

Since cases 2-4 have a minimum < case 1, I don't really need to precisely calculate the finalization gap.

tl;dr
I believe the maximum safe lookahead is 4.
I'll make a PR :)

**Comment by michaelsproul** [comment]
Closed by #1300
